extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -308,19 +308,22 @@ public void stop(boolean stopGracefully)
       // until the tasks have acknowledged or timed out. We want this behavior when we're explicitly shut down through
       // the API, but if we shut down for other reasons (e.g. we lose leadership) we want to just stop and leave the
       // tasks as they are.
-      if (stopGracefully) {
-        log.info("Stopping gracefully, signalling managed tasks to complete and publish");
-        synchronized (stopLock) {
+      synchronized (stopLock) {
+        if (stopGracefully) {
+          log.info("Posting GracefulShutdownNotice, signalling managed tasks to complete and publish");
+          notices.add(new GracefulShutdownNotice());
+        } else {
+          log.info("Posting ShutdownNotice");
           notices.add(new ShutdownNotice());
-          while (!stopped) {
-            stopLock.wait();
-          }
         }
-        log.info("Shutdown notice handled");
+
+        while (!stopped) {
+          stopLock.wait();
```

Comment:
Sorry, I meant, since this is called by a jetty thread, it's kind of dangerous to have an infinite timeout. That could potentially lock a jetty thread forever. So I was thinking of adding a timeout, but then wanted to understand the behavior there.

IOW: would anything bad happen on the overlord if we added a timeout, such that this stop method returns (by throwing an exception) but the supervisor run thread is still going? If not- let's add a timeout.